File: togo/internal/services/type_hierarchy_test.go
package services

import (
	"strings"
	"testing"

	"github.com/nikeee/cappu/internal/compiler"
	"github.com/nikeee/cappu/internal/lsp"
)

// Port of src/services/typeHierarchy.test.ts.

const thSrc = "interface Shape {}\n" +
	"class Base implements Shape {}\n" +
	"class Mid extends Base {}\n" +
	"class Leaf extends Mid {}"

func thSetup(text string) (*compiler.Program, *compiler.Checker, *compiler.Node) {
	program := compiler.NewProgram()
	compiler.LoadJdkStub(program)
	program.SetOpenDocument("file:///S.java", text, 1)
	return program, compiler.NewChecker(program), program.GetSourceFile("file:///S.java")
}

func itemNames(items []lsp.TypeHierarchyItem) []string {
	out := []string{}
	for _, i := range items {
		out = append(out, i.Name)
	}
	return out
}

func TestPrepareTypeHierarchy(t *testing.T) {
	_, checker, sf := thSetup(thSrc)
	items := PrepareTypeHierarchy(checker, sf, strings.Index(thSrc, "Mid extends"))
	if got := itemNames(items); len(got) != 1 || got[0] != "Mid" {
		t.Fatalf("prepare = %v, want [Mid]", got)
	}
}

func TestTypeHierarchyNeighbours(t *testing.T) {
	program, checker, sf := thSetup(thSrc)
	mid := PrepareTypeHierarchy(checker, sf, strings.Index(thSrc, "Mid extends"))[0]
	if got := itemNames(TypeHierarchySupertypes(program, checker, mid)); len(got) != 1 || got[0] != "Base" {
		t.Errorf("supertypes = %v, want [Base]", got)
	}
	if got := itemNames(TypeHierarchySubtypes(program, checker, mid)); len(got) != 1 || got[0] != "Leaf" {
		t.Errorf("subtypes = %v, want [Leaf]", got)
	}
}

func TestTypeHierarchySupertypesIncludesInterface(t *testing.T) {
	program, checker, sf := thSetup(thSrc)
	base := PrepareTypeHierarchy(checker, sf, strings.Index(thSrc, "Base implements"))[0]
	found := false
	for _, n := range itemNames(TypeHierarchySupertypes(program, checker, base)) {
		if n == "Shape" {
			found = true
		}
	}
	if !found {
		t.Error("Base's supertypes should include Shape")
	}
}

func TestPrepareTypeHierarchyNotAType(t *testing.T) {
	_, checker, sf := thSetup(thSrc)
	if items := PrepareTypeHierarchy(checker, sf, 0); len(items) != 0 {
		t.Errorf("prepare at non-type = %v, want empty", itemNames(items))
	}
}

File: src/services/typeHierarchy.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { createChecker } from "../compiler/checker.ts";
import { loadJdkStub } from "../compiler/jdkStub.ts";
import { createProgram } from "../compiler/program.ts";
import { type Uri } from "../workspace.ts";
import {
  prepareTypeHierarchy,
  typeHierarchySubtypes,
  typeHierarchySupertypes,
} from "./typeHierarchy.ts";

const SRC = [
  "interface Shape {}",
  "class Base implements Shape {}",
  "class Mid extends Base {}",
  "class Leaf extends Mid {}",
].join("\n");

function setup(text: string) {
  const program = createProgram();
  loadJdkStub(program);
  program.setOpenDocument("file:///S.java" as Uri, text, 1);
  return {
    program,
    checker: createChecker(program),
    sourceFile: program.getSourceFile("file:///S.java" as Uri)!,
  };
}

test("prepare resolves the type at the cursor", () => {
  const { checker, sourceFile } = setup(SRC);
  const items = prepareTypeHierarchy(checker, sourceFile, SRC.indexOf("Mid extends"));
  expect(items?.map(i => i.name)).toEqual(["Mid"]);
});

test("supertypes and subtypes return the direct neighbours", () => {
  const { program, checker, sourceFile } = setup(SRC);
  const [mid] = prepareTypeHierarchy(checker, sourceFile, SRC.indexOf("Mid extends"))!;
  expect(typeHierarchySupertypes(program, checker, mid!)?.map(i => i.name)).toEqual(["Base"]);
  expect(typeHierarchySubtypes(program, checker, mid!)?.map(i => i.name)).toEqual(["Leaf"]);
});

test("supertypes of an interface implementor includes the interface", () => {
  const { program, checker, sourceFile } = setup(SRC);
  const [base] = prepareTypeHierarchy(checker, sourceFile, SRC.indexOf("Base implements"))!;
  // Base implements Shape (and implicitly extends Object).
  expect(typeHierarchySupertypes(program, checker, base!)?.map(i => i.name)).toContain("Shape");
});

test("prepare returns null when the cursor is not on a type", () => {
  const { checker, sourceFile } = setup(SRC);
  expect(prepareTypeHierarchy(checker, sourceFile, 0)).toBeNull();
});
